File: src/MonkeyLang.Test/LexerTests.cs
using System;
using System.Collections.Generic;
using System.Text;
using FluentAssertions;
using MonkeyLang.Lexical;
using Xunit;

namespace MonkeyLang.Test
{
    public class LexerTests
    {

        [Theory]
        [InlineData("let x = y;", 'l')]
        [InlineData("=+(){},;", '=')]
        [InlineData("fn", 'f')]
        public void Test_LexerInitializesCorrectly(string input, char firstChar)
        {
            var lexer = new Lexer(input);

            Assert.Equal(input, lexer.input);
            Assert.Equal(0, lexer.position);
            Assert.Equal(1, lexer.readPosition);
            Assert.Equal(firstChar, lexer.ch);
        }

        [Theory]
        [InlineData("let x = y;", 't', 2, 3)]
        [InlineData("=+(){},;", '(', 2, 3)]
        [InlineData("fn", '\0', 1, 2)]
        public void Test_ReadChar(string input, char expectedChar, int expectedPosition, int expectedReadPosition)
        {
            var lexer = new Lexer(input);

            lexer.readChar();
            lexer.readChar();

            lexer.ch.Should().Be(expectedChar);
            lexer.position.Should().Be(expectedPosition);
            lexer.readPosition.Should().Be(expectedReadPosition);
        }

        [Theory]
        [InlineData("let x = y;", ' ')]
        [InlineData("=+(){},;", ')')]
        [InlineData("fn", '\0')]
        public void Test_PeekChar(string input, char expectedPeek)
        {
            var lexer = new Lexer(input);

            lexer.readChar();
            lexer.readChar();

            var peekChar = lexer.peekChar();

            peekChar.Should().Be(expectedPeek);
        }

        [Theory]
        [InlineData('a', true)]
        [InlineData('m', true)]
        [InlineData('z', true)]
        [InlineData('A', true)]
        [InlineData('J', true)]
        [InlineData('Z', true)]
        [InlineData('_', true)]
        [InlineData('@', false)]
        [InlineData('[', false)]
        [InlineData('0', false)]
        [InlineData('2', false)]
        [InlineData('{', false)]
        public void Test_IsLetter(char ch, bool expected)
        {
            var isLetter = Lexer.isLetter(ch);

            isLetter.Should().Be(expected);
        }

        [Fact]
        public void Test_NextToken()
        {
            string input = "=+(){},;";

            var expectedTokens = new List<Token>()
            {
                new Token(TokenType.ASSIGN, "="),
                new Token(TokenType.PLUS, "+"),
                new Token(TokenType.LPAREN, "("),
                new Token(TokenType.RPAREN, ")"),
                new Token(TokenType.LBRACE, "{"),
                new Token(TokenType.RBRACE, "}"),
                new Token(TokenType.COMMA, ","),
                new Token(TokenType.SEMICOLON, ";"),
                new Token(TokenType.EOF, "")
            };

            var lexer = new Lexer(input);
            foreach (var expectedToken in expectedTokens)
            {
                var nextToken = lexer.NextToken();
                Assert.Equal(expectedToken.Type, nextToken.Type);
                Assert.Equal(expectedToken.Literal, nextToken.Literal);
            }
        }

        [Fact]
        public void Test_Small_Program()
        {
            string input = @"let five = 5;
let ten = 10;

let add = fn(x, y) {
    x + y;
};

let result = add(five, ten);
";
            var expectedTokens = new List<Token>()
            {
                new Token(TokenType.LET, "let"),
                new Token(TokenType.IDENT, "five"),
                new Token(TokenType.ASSIGN, "="),
                new Token(TokenType.INT, "5"),
                new Token(TokenType.SEMICOLON, ";"),

                new Token(TokenType.LET, "let"),
                new Token(TokenType.IDENT, "ten"),
                new Token(TokenType.ASSIGN, "="),
                new Token(TokenType.INT, "10"),
                new Token(TokenType.SEMICOLON, ";"),

                new Token(TokenType.LET, "let"),
                new Token(TokenType.IDENT, "add"),
                new Token(TokenType.ASSIGN, "="),
                new Token(TokenType.FUNCTION, "fn"),
                new Token(TokenType.LPAREN, "("),
                new Token(TokenType.IDENT, "x"),
                new Token(TokenType.COMMA, ","),
                new Token(TokenType.IDENT, "y"),
                new Token(TokenType.RPAREN, ")"),
                new Token(TokenType.LBRACE, "{"),
                new Token(TokenType.IDENT, "x"),
                new Token(TokenType.PLUS, "+"),
                new Token(TokenType.IDENT, "y"),
                new Token(TokenType.SEMICOLON, ";"),
                new Token(TokenType.RBRACE, "}"),
                new Token(TokenType.SEMICOLON, ";"),

                new Token(TokenType.LET, "let"),
                new Token(TokenType.IDENT, "result"),
                new Token(TokenType.ASSIGN, "="),
                new Token(TokenType.IDENT, "add"),
                new Token(TokenType.LPAREN, "("),
                new Token(TokenType.IDENT, "five"),
                new Token(TokenType.COMMA, ","),
                new Token(TokenType.IDENT, "ten"),
                new Token(TokenType.RPAREN, ")"),
                new Token(TokenType.SEMICOLON, ";"),
                new Token(TokenType.EOF, "")

            };

            var lexer = new Lexer(input);
            foreach (var expectedToken in expectedTokens)
            {
                var nextToken = lexer.NextToken();
                Assert.Equal(expectedToken.Type, nextToken.Type);
                Assert.Equal(expectedToken.Literal, nextToken.Literal);
            }
        }

        [Fact]
        public void Test_More_Operators()
        {
            var input = @"
!-/*5;
5 < 10 > 5;
";

            var expectedTokens = new List<Token>()
            {
                new Token(TokenType.BANG, "!"),
                new Token(TokenType.MINUS, "-"),
                new Token(TokenType.SLASH, "/"),
                new Token(TokenType.ASTERISK, "*"),
                new Token(TokenType.INT, "5"),
                new Token(TokenType.SEMICOLON, ";"),
                new Token(TokenType.INT, "5"),
                new Token(TokenType.LT, "<"),
                new Token(TokenType.INT, "10"),
                new Token(TokenType.GT, ">"),
                new Token(TokenType.INT, "5"),
                new Token(TokenType.SEMICOLON, ";"),
                new Token(TokenType.EOF, "")
            };

            var lexer = new Lexer(input);
            foreach (var expectedToken in expectedTokens)
            {
                var nextToken = lexer.NextToken();
                Assert.Equal(expectedToken.Type, nextToken.Type);
                Assert.Equal(expectedToken.Literal, nextToken.Literal);
            }
        }

        [Fact]
        public void Test_Conditional_Items()
        {
            var input = @"
if (5 < 10) {
    return true;
} else {
    return false;
}";

            var expectedTokens = new List<Token>()
            {
                new Token(TokenType.IF, "if"),
                new Token(TokenType.LPAREN, "("),
                new Token(TokenType.INT, "5"),
                new Token(TokenType.LT, "<"),
                new Token(TokenType.INT, "10"),
                new Token(TokenType.RPAREN, ")"),
                new Token(TokenType.LBRACE, "{"),
                new Token(TokenType.RETURN, "return"),
                new Token(TokenType.TRUE, "true"),
                new Token(TokenType.SEMICOLON, ";"),

                new Token(TokenType.RBRACE, "}"),
                new Token(TokenType.ELSE, "else"),
                new Token(TokenType.LBRACE, "{"),


                new Token(TokenType.RETURN, "return"),
                new Token(TokenType.FALSE, "false"),
                new Token(TokenType.SEMICOLON, ";"),

                new Token(TokenType.RBRACE, "}"),
                new Token(TokenType.EOF, "")
            };

            var lexer = new Lexer(input);
            foreach (var expectedToken in expectedTokens)
            {
                var nextToken = lexer.NextToken();
                Assert.Equal(expectedToken.Type, nextToken.Type);
                Assert.Equal(expectedToken.Literal, nextToken.Literal);
            }
        }

        [Fact]
        public void Test_Double_Comparisons()
        {
            var input = @"
10 == 10;
10 != 9;
";
            var expectedTokens = new List<Token>()
            {
                new Token(TokenType.INT, "10"),
                new Token(TokenType.EQ, "=="),
                new Token(TokenType.INT, "10"),
                new Token(TokenType.SEMICOLON, ";"),

                new Token(TokenType.INT, "10"),
                new Token(TokenType.NOT_EQ, "!="),
                new Token(TokenType.INT, "9"),
                new Token(TokenType.SEMICOLON, ";"),
                new Token(TokenType.EOF, "")
            };

            var lexer = new Lexer(input);
            foreach (var expectedToken in expectedTokens)
            {
                var nextToken = lexer.NextToken();
                Assert.Equal(expectedToken.Type, nextToken.Type);
                Assert.Equal(expectedToken.Literal, nextToken.Literal);
            }

        }
    }
}

File: src/MonkeyLang.Tests/LexerTests.fs
module LexerTests

open System
open Xunit
open Tokens
open Lexer

let AssertTokens(lexer: LexerState, expectedToken) =
    let actualToken = nextToken lexer
    Assert.Equal(expectedToken.TokenType, actualToken.TokenType)
    Assert.Equal(expectedToken.Literal, actualToken.Literal)
    ()

let buildTokenTypes(tokens) =
    tokens |> List.map (fun (t, l) -> { TokenType = t ; Literal = l })

[<Fact>]
let ``Can Lex Basic Symbols`` () =
    let expectedTokens =
        [
            { TokenType = TokenType.ASSIGN ; Literal = "=" };
            { TokenType = TokenType.PLUS ; Literal = "+" };
            { TokenType = TokenType.LPAREN ; Literal = "(" };
            { TokenType = TokenType.RPAREN ; Literal = ")" };
            { TokenType = TokenType.LBRACE ; Literal = "{" };
            { TokenType = TokenType.RBRACE ; Literal = "}" };
            { TokenType = TokenType.COMMA ; Literal = "," };
            { TokenType = TokenType.SEMICOLON ; Literal = ";" };
            { TokenType = TokenType.EOF ; Literal = "" };
        ]
    
    let input = "=+(){},;"
    let lexer = createLexer input
    expectedTokens |> List.iter (fun et -> AssertTokens(lexer, et))

[<Fact>]
let ``Can Lex Identifiers`` () =
    let input = "let five = 5;
let ten = 10;

let add = fn(x, y) {
  x + y;
};

let result = add(five, ten);"
    
    let expectedTokensRaw:(TokenType * string) list =
        [
            (TokenType.LET, "let");
            (TokenType.IDENT, "five");
            (TokenType.ASSIGN, "=");
            (TokenType.INT, "5");
            (TokenType.SEMICOLON, ";");
            (TokenType.LET, "let");
            (TokenType.IDENT, "ten");
            (TokenType.ASSIGN, "=");
            (TokenType.INT, "10");
            (TokenType.SEMICOLON, ";");
            (TokenType.LET, "let");
            (TokenType.IDENT, "add");
            (TokenType.ASSIGN, "=");
            (TokenType.FUNCTION, "fn");
            (TokenType.LPAREN, "(");
            (TokenType.IDENT, "x");
            (TokenType.COMMA, ",");
            (TokenType.IDENT, "y");
            (TokenType.RPAREN, ")");
            (TokenType.LBRACE, "{");
            (TokenType.IDENT, "x");
            (TokenType.PLUS, "+");
            (TokenType.IDENT, "y");
            (TokenType.SEMICOLON, ";");
            (TokenType.RBRACE, "}");
            (TokenType.SEMICOLON, ";");
            (TokenType.LET, "let");
            (TokenType.IDENT, "result");
            (TokenType.ASSIGN, "=");
            (TokenType.IDENT, "add");
            (TokenType.LPAREN, "(");
            (TokenType.IDENT, "five");
            (TokenType.COMMA, ",");
            (TokenType.IDENT, "ten");
            (TokenType.RPAREN, ")");
            (TokenType.SEMICOLON, ";");
            (TokenType.EOF, "");
        ]

    let expectedTokens = buildTokenTypes expectedTokensRaw
    let lexer = createLexer input
    expectedTokens |> List.iter (fun et -> AssertTokens(lexer, et))

[<Fact>]
let ``Can Lex More Tokens`` () =
    let input = "!-/*5;

5 < 10 > 5;"
    
    let expectedTokensRaw:(TokenType * string) list =
        [
            (TokenType.BANG, "!");
            (TokenType.MINUS, "-");
            (TokenType.SLASH, "/");
            (TokenType.ASTERISK, "*");
            (TokenType.INT, "5");
            (TokenType.SEMICOLON, ";");
            (TokenType.INT, "5");
            (TokenType.LT, "<");
            (TokenType.INT, "10");
            (TokenType.GT, ">");
            (TokenType.INT, "5");
            (TokenType.SEMICOLON, ";");
            (TokenType.EOF, "");
        ]

    let expectedTokens = buildTokenTypes expectedTokensRaw
    let lexer = createLexer input
    expectedTokens |> List.iter (fun et -> AssertTokens(lexer, et))

[<Fact>]
let ``Can Lex more keywords`` () =
    let input = "if (5 < 10) {
        return true;
    } else {
        return false;
    }"

    let expectedTokensRaw:(TokenType * string) list =
        [

            (TokenType.IF, "if");
            (TokenType.LPAREN, "(");
            (TokenType.INT, "5");
            (TokenType.LT, "<");
            (TokenType.INT, "10");
            (TokenType.RPAREN, ")");
            (TokenType.LBRACE, "{");
            (TokenType.RETURN, "return");
            (TokenType.TRUE, "true");
            (TokenType.SEMICOLON, ";");
            (TokenType.RBRACE, "}");
            (TokenType.ELSE, "else");
            (TokenType.LBRACE, "{");
            (TokenType.RETURN, "return");
            (TokenType.FALSE, "false");
            (TokenType.SEMICOLON, ";");
            (TokenType.RBRACE, "}");
            (TokenType.EOF, "");
        ]

    let expectedTokens = buildTokenTypes expectedTokensRaw
    let lexer = createLexer input
    expectedTokens |> List.iter (fun et -> AssertTokens(lexer, et))
    
[<Fact>]
let ``Can Lex advanced equals/not equals operators`` () =
    let input = "10 == 10
10 != 9"

    let expectedTokensRaw:(TokenType * string) list =
        [
            (TokenType.INT, "10");
            (TokenType.EQ, "==");
            (TokenType.INT, "10");
            (TokenType.INT, "10");
            (TokenType.NOT_EQ, "!=");
            (TokenType.INT, "9");
            (TokenType.EOF, "");
        ]

    let expectedTokens = buildTokenTypes expectedTokensRaw
    let lexer = createLexer input
    expectedTokens |> List.iter (fun et -> AssertTokens(lexer, et))

[<Fact>]
let ``Can lex strings`` () =
    let input = "\"foobar\"
    \"foo bar\""

    let expectedTokensRaw:(TokenType * string) list =
        [
            (TokenType.STRING, "foobar");
            (TokenType.STRING, "foo bar");
            (TokenType.EOF, "");
        ]

    let expectedTokens = buildTokenTypes expectedTokensRaw
    let lexer = createLexer input
    expectedTokens |> List.iter (fun et -> AssertTokens(lexer, et))

[<Fact>]
let ``Can lex array`` () =
    let input = "[1, 2];"

    let expectedTokensRaw:(TokenType * string) list =
        [
            (TokenType.LBRACKET, "[");
            (TokenType.INT, "1");
            (TokenType.COMMA, ",");
            (TokenType.INT, "2");
            (TokenType.RBRACKET, "]");
            (TokenType.SEMICOLON, ";");
            (TokenType.EOF, "");
        ]

    let expectedTokens = buildTokenTypes expectedTokensRaw
    let lexer = createLexer input
    expectedTokens |> List.iter (fun et -> AssertTokens(lexer, et))
[<Fact>]
let ``Can lex hash`` () =
    let input = "{ 1: \"23\"};"

    let expectedTokensRaw:(TokenType * string) list =
        [
            (TokenType.LBRACE, "{");
            (TokenType.INT, "1");
            (TokenType.COLON, ":");
            (TokenType.STRING, "23");
            (TokenType.RBRACE, "}");
            (TokenType.SEMICOLON, ";");
            (TokenType.EOF, "");
        ]

    let expectedTokens = buildTokenTypes expectedTokensRaw
    let lexer = createLexer input
    expectedTokens |> List.iter (fun et -> AssertTokens(lexer, et))